src/inmanta/server/services/compilerservice.py

```diff
@@ -1095,11 +1094,13 @@ async def cancel_compile(self, env_id: uuid.UUID) -> None:
 
         :param env_id: The id of the environment for which the compile should be cancelled.
         """
-        compile_task: Optional[Task[None | Result]] = self._recompiles.pop(env_id, None)
-        if compile_task:
-            compile_task.cancel()
-            try:
-                # Wait until cancellation has finished
-                await compile_task
-            except asyncio.CancelledError:
-                pass
+        async with self._write_lock_env_to_compile_task:
+            compile_task: Optional[asyncio.Task[None | protocol.Result]] = self._env_to_compile_task.pop(env_id, None)
+            if compile_task:
+                compile_task.cancel()
+                try:
+                    # Wait until cancellation has finished
+                    await compile_task
+                except asyncio.CancelledError:
+                    # Don't propagate CancelledError raised by invoking compile_task.cancel()
+                    pass
```

Comment:
I think we want to `pop` the task within the lock, but waiting for it to complete should be done outside of it to keep the lock shortlived. After all, we don't want to block other environments.

This is only safe if the environment is halted when we're here, but I believe this is the case?